src/Mod/Part/App/PrimitiveFeature.cpp

```diff
@@ -940,12 +942,38 @@ App::DocumentObjectExecReturn *Spiral::execute(void)
         gp_Pnt2d cend(u, v);
         end = cend;
 
-        Handle(Geom2d_TrimmedCurve) segm = GCE2d_MakeSegment(beg , end);
+//         Handle(Geom2d_TrimmedCurve) segm = GCE2d_MakeSegment(beg , end);
+// 
+//         TopoDS_Edge edgeOnSurf = BRepBuilderAPI_MakeEdge(segm , surf);
+//         TopoDS_Wire wire = BRepBuilderAPI_MakeWire(edgeOnSurf);
+//         BRepLib::BuildCurves3d(wire);
 
-        TopoDS_Edge edgeOnSurf = BRepBuilderAPI_MakeEdge(segm , surf);
-        TopoDS_Wire wire = BRepBuilderAPI_MakeWire(edgeOnSurf);
-        BRepLib::BuildCurves3d(wire);
+        BRepBuilderAPI_MakeWire mkWire;
+        Handle(Geom2d_TrimmedCurve) segm;
+        TopoDS_Edge edgeOnSurf;
+
+        for (unsigned long i = 0; i < floor(myNumRot); i++) {
+            u = coneDir * (i+1) * 2.0 * M_PI;
+            v = ((i+1) * myPitch) / cos(myAngle);
+            end = gp_Pnt2d(u, v);
+            segm = GCE2d_MakeSegment(beg , end);
+            edgeOnSurf = BRepBuilderAPI_MakeEdge(segm , surf);
+            mkWire.Add(edgeOnSurf);
+            beg = end;
+        }
 
+        if (partTurn > Precision::Confusion()) {
+            u = coneDir * myNumRot * 2.0 * M_PI;
+            v = myHeight / cos(myAngle);
+            end = gp_Pnt2d(u, v);
+            segm = GCE2d_MakeSegment(beg , end);
+            edgeOnSurf = BRepBuilderAPI_MakeEdge(segm , surf);
+            mkWire.Add(edgeOnSurf);
+        }
+
+        TopoDS_Wire wire = mkWire.Wire();
+        BRepLib::BuildCurves3d(wire);

```

Comment:
You might consider writing a helper function, either inside `execute` itself or as a private member function of `Wedge`. This will allow you to simplify the code in `execute` a bit, making it more legible. It also (more importantly) reduces code duplication.

here is a stab at what this helper function could look like.

```cpp
/** Adds a single segment to the Spiral being built
 *
 *  Note that the `gp_Pnt2d` and the `BRepBuiderAPI_MakeWire` references passed
 *  in are both modified, therefore they are not `const` in the argument list
 */
void Wedge::addSpiralSegment(float u, float v, gp_Pnt2d& beg, BRepBuilderAPI_MakeWire& builder)
{
    // First, create the new end-point using the provided values
    gp_Pnt2d end = gp_Pnt2d(u, v);
    // Use the provided beginning-point and the created end-point to create the
    // new segment
    Handle(Geom2d_TrimmedCurve) segm = GCE2d_MakeSegment(beg , end);
    // Create a topological Edge out the segment
    TopoDS_Edge edgeOnSurf = BRepBuilderAPI_MakeEdge(segm , surf);
    // Finally, add the segment to our  builder
    builder.Add(edgeOnSurf);
    // Update the beggining-point for the next iteration (if any)
    beg = end;
}
```